Added options to readability-implicit-bool-conversion

4m4n-x-B4w4ne opened the pull request:
Fixes: #36323 , I added two new options in the clang-tools-extra/clan-tidy/readibility/ImplicitBoolConversionCheck.cpp and header file.
I have also written new test cases to test these new options in test/readibility directory.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Added new options in ImplicitBoolConversionCheck CheckConversionToBool and CheckConversionFromBool.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.h

Added CheckConversionToBool and CheckConversionFromBool Options in the header

4m4n-x-B4w4ne committed:
Create implicit-bool-conversion-check.cpp

Added new test to check the new options added in the ImplicitBoolConversionCheck.cpp

4m4n-x-B4w4ne committed:
Merge branch 'main' into main

EugeneZelenko commented:
Please mention options in Release Notes and check documentation.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Fixed Bugs

4m4n-x-B4w4ne committed:
Update implicit-bool-conversion-check.cpp

Fixed Bugs in test .

HerrCai0907 reviewed (commented):
update release note and doc also

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:
```diff
@@ -0,0 +1,86 @@
+// RUN: %check_clang_tidy %s readability-implicit-bool-conversion %t
+
+// RUN: %check_clang_tidy %s readability-implicit-bool-conversion %t \
+// RUN: -- -config='{CheckOptions: [{key: readability-implicit-bool-conversion.CheckConversionsToBool, value: false}, {key: readability-implicit-bool-conversion.CheckConversionsFromBool, value: true}]}'
```
this is old format for options, use new one like in other checks

4m4n-x-B4w4ne replied:
Okay Thanks I am Working on it.

4m4n-x-B4w4ne replied:
I have done this . Can you please review it?

PiotrZSL commented on clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:
```diff
@@ -357,20 +360,22 @@ void ImplicitBoolConversionCheck::registerMatchers(MatchFinder *Finder) {
 void ImplicitBoolConversionCheck::check(
     const MatchFinder::MatchResult &Result) {
-
+  if(CheckConversionsToBool){
```
Wrong, place these things should be actually put into matcher in registerMatchers method, to speed up.

You got 2 calls to addMatcher, one in line 300, other in line 339, put those (with dependences to avoid unused local variables) into if's.

4m4n-x-B4w4ne replied:
Okay Thanks I am Working on it.

4m4n-x-B4w4ne replied:
Can You please check this? I have done this now as instructed.

PiotrZSL reviewed (changes requested):
- reformat code
- missing release notes entry about new options added
- missing documentation for newly added options
- change should be in registerMatchers method

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

Added new Options in ImplicitBoolConversionCheck.

4m4n-x-B4w4ne committed:
Merge branch 'llvm:main' into main

4m4n-x-B4w4ne commented:
> update release note and doc also

I am doing it.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -332,6 +332,11 @@ Changes in existing checks
   <clang-tidy/checks/readability/identifier-naming>` check to
   validate ``namespace`` aliases.
 
+-Improved :doc: `readability-implicit-bool-conversion
```
Please extend existing `readability-implicit-bool-conversion` entry.

4m4n-x-B4w4ne replied:
Done Thanks.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -332,6 +332,11 @@ Changes in existing checks
 
+-Improved :doc: `readability-implicit-bool-conversion
+ <clang-tidy/checks/readability/implicit-bool-conversion>`
+  - `CheckConversionsToBool`: Allows enabling or disabling warnings for implicit conversions to `bool` (default) : true .
```
Default belongs to documentation.

4m4n-x-B4w4ne replied:
Done Thanks.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,10 @@ Changes in existing checks
-  case of the literal suffix in fixes and fixing false positive for implicit
-  conversion of comparison result in C23.
+  <clang-tidy/checks/readability/implicit-bool-conversion>` 
+   - `UseUpperCaseLiteralSuffix` : check by adding the option to select the case of the literal suffix in fixes and fixing false positive for implicit conversion of comparison result in C23.
```
`check by adding the options` should be common for all three options.

EugeneZelenko replied:
Sorry, if I was not clear, `check by adding the options` should be added once before list of added options.

4m4n-x-B4w4ne replied:
Thanks I have done it in Release Notes .

EugeneZelenko replied:
It is not necessary to repeat `check by adding the option` thrice.

4m4n-x-B4w4ne replied:
Its looking good ?

EugeneZelenko commented:
New options are still not described in check documentation.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -320,9 +320,10 @@ Changes in existing checks
-   - `UseUpperCaseLiteralSuffix` : check by adding the option to select the case of the literal suffix in fixes and fixing false positive for implicit conversion of comparison result in C23.
-   - `CheckConversionsToBool`: check by adding the option to allow enabling or disabling warnings for implicit conversions to `bool` .
-   - `CheckConversionsFromBool`: check by adding the option to allow enabling or disabling warnings for implicit conversions from `bool`.
+   Check by adding the option 
```
Belongs to previous line.

4m4n-x-B4w4ne replied:
Sorry but i havent got this one . Can you please clarify?

4m4n-x-B4w4ne replied:
Thanks got it now.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -320,9 +320,10 @@ Changes in existing checks
-   - `CheckConversionsToBool`: check by adding the option to allow enabling or disabling warnings for implicit conversions to `bool` .
-   - `CheckConversionsFromBool`: check by adding the option to allow enabling or disabling warnings for implicit conversions from `bool`.
+   Check by adding the option 
+   - `UseUpperCaseLiteralSuffix` : to select the case of the literal suffix in fixes and fixing false positive for implicit conversion of comparison result in C23.
```
Please follow 80-characters limit.

4m4n-x-B4w4ne replied:
Done Thanks.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,15 @@ Changes in existing checks
-  case of the literal suffix in fixes and fixing false positive for implicit
-  conversion of comparison result in C23.
+  <clang-tidy/checks/readability/implicit-bool-conversion>` check by adding 
+   the option 
```
```suggestion
   the options:
```

Or `next options:`?

4m4n-x-B4w4ne replied:
Done Thanks.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,15 @@ Changes in existing checks
-  conversion of comparison result in C23.
+  <clang-tidy/checks/readability/implicit-bool-conversion>` check by adding 
+   the option 
+   - `UseUpperCaseLiteralSuffix` : to select the case of the literal suffix in 
```
```suggestion
   - `UseUpperCaseLiteralSuffix`: to select the case of the literal suffix in 
```

4m4n-x-B4w4ne replied:
Done Thanks.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,15 @@ Changes in existing checks
+      fixes and fixing false positive for implicit conversion of comparison 
+      result in C23.
+   - `CheckConversionsToBool`: to allow enabling or disabling warnings for 
+      implicit conversions to `bool` .
```
```suggestion
      implicit conversions to ``bool``.
```

4m4n-x-B4w4ne replied:
Done Thanks.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,15 @@ Changes in existing checks
+   - `CheckConversionsToBool`: to allow enabling or disabling warnings for 
+      implicit conversions to `bool` .
+   - `CheckConversionsFromBool`: to allow enabling or disabling warnings for 
+      implicit conversions from `bool`.
```
```suggestion
      implicit conversions from ``bool``.
```

4m4n-x-B4w4ne replied:
Done Thanks.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,15 @@ Changes in existing checks
+      fixes and fixing false positive for implicit conversion of comparison 
+      result in C23.
+   - `CheckConversionsToBool`: to allow enabling or disabling warnings for 
+      implicit conversions to ``bool`` .
```
```suggestion
      implicit conversions to ``bool``.
```

4m4n-x-B4w4ne replied:
Done Thanks.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

Improved Preview By adding blank line.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Fixed Bugs .

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Fixed Bugs .

EugeneZelenko commented on clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:
```diff
@@ -277,23 +281,18 @@ void ImplicitBoolConversionCheck::registerMatchers(MatchFinder *Finder) {
+  auto IsInCompilerGeneratedFunction = hasAncestor(namedDecl(anyOf(
+      isImplicit(), functionDecl(isDefaulted()), functionTemplateDecl())));
+    
+  if(CheckConversionToBool){
```
Please run Clang-format.

4m4n-x-B4w4ne replied:
Done Clang-Format .
Thanks.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Fixed Bugs.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Formatted using clang-format.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Fixed Some Bugs.

4m4n-x-B4w4ne committed:
Update implicit-bool-conversion.rst

Documented the new options.

4m4n-x-B4w4ne commented:
> Please mention options in Release Notes and check documentation.

I have documented it can you please review it !

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/readability/implicit-bool-conversion.rst:
```diff
@@ -147,3 +147,38 @@ Options
       // ^ propose replacement with option `UseUpperCaseLiteralSuffix`: if (foo != 0U) {}
+.. option:: CheckConversionsToBool
+
+   When `true`, the check diagnoses implicit conversions to `bool`. Default is `true`.
```
```suggestion
   When `true`, the check diagnoses implicit conversions to ``bool``. Default is `true`.
```

As language construct. Please also follow 80 characters limit. Same for other option.

4m4n-x-B4w4ne replied:
Done Thanks.

Can you please also review ImplicitBoolConversionCheck.cpp?

EugeneZelenko replied:
Please wait for other reviewers. I deal only with documentation and minor code style issues.

4m4n-x-B4w4ne replied:
Okay Thanks a lot !

4m4n-x-B4w4ne committed:
Update implicit-bool-conversion.rst

4m4n-x-B4w4ne committed:
Update implicit-bool-conversion-check.cpp

Rewritten the RUN directives.

PiotrZSL reviewed (approved):
Except pointed out findings, looks ok.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

Fixed by separating options by ',' .

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

4m4n-x-B4w4ne commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,16 @@ Changes in existing checks
+  <clang-tidy/checks/readability/implicit-bool-conversion>` check by adding the
+   options:
+
+   - `UseUpperCaseLiteralSuffix`: to select the case of the literal suffix in 
```
Done Thanks.

4m4n-x-B4w4ne commented on clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:
```diff
@@ -259,100 +259,118 @@ ImplicitBoolConversionCheck::ImplicitBoolConversionCheck(
-              unless(isInTemplateInstantiation()),
-              unless(IsInCompilerGeneratedFunction))),
-      this);
+  if (CheckConversionFromBool || CheckConversionToBool) {
```
Done Thanks.

4m4n-x-B4w4ne commented:
Hey @EugeneZelenko  , As everything got reviewed .Is it possible to merge my PR ?

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,9 @@ Changes in existing checks
-  conversion of comparison result in C23.
+  <clang-tidy/checks/readability/implicit-bool-conversion>` check by adding the
+   options:
+   `UseUpperCaseLiteralSuffix`,`CheckConversionsToBool`,`CheckConversionsFromBool`.
```
Please merge with previous line.

4m4n-x-B4w4ne replied:
Done Thanks.

4m4n-x-B4w4ne replied:
Hey what needs to be done further!

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

5chmidti commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,9 @@ Changes in existing checks
-  <clang-tidy/checks/readability/implicit-bool-conversion>` check
-  by adding the option `UseUpperCaseLiteralSuffix` to select the
-  case of the literal suffix in fixes and fixing false positive for implicit
-  conversion of comparison result in C23.
+  <clang-tidy/checks/readability/implicit-bool-conversion>` check by adding the
+   options: `UseUpperCaseLiteralSuffix`,`CheckConversionsToBool`,
+   `CheckConversionsFromBool`.
```
Please re-add what was already there to end up with something like: 

```
- Improved :doc:`readability-implicit-bool-conversion
  <clang-tidy/checks/readability/implicit-bool-conversion>` check
  by adding the option `UseUpperCaseLiteralSuffix` to select the
  case of the literal suffix in fixes and fixing false positive for implicit
  conversion of comparison result in C23, and by adding the
  options `CheckConversionsToBool` and
  `CheckConversionsFromBool` to configure which conversions 
  involving ``bool`` are allowed.
```

4m4n-x-B4w4ne replied:
Done Thanks.

5chmidti commented on clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:
```diff
@@ -287,72 +292,83 @@ void ImplicitBoolConversionCheck::registerMatchers(MatchFinder *Finder) {
-              unless(isInTemplateInstantiation()),
-              unless(IsInCompilerGeneratedFunction))),
-      this);
+  if (CheckConversionToBool) {
```
The PR doesn't build because you have a typo: `CheckConversionToBool` -> `CheckConversionsToBool`, the same applies to `CheckConversionFromBool` below.

4m4n-x-B4w4ne replied:
Done Thanks. Sorry for my carelessness.

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:
You have added `-check-suffix` options to each of your `RUN` lines, but none of your `CHECK-MESSAGES` are actually using it. E.g.,: 
```C++
// CHECK-MESSAGES-TO-BOOL-FALSE
```

Although, instead of using the suffixes to say `FALSE`, you could swap them around to say that the option is enabled, so `-check-suffix:TO-BOOL`. Then, the version with both enabled becomes `-check-suffix:TO-BOOL,FROM-BOOL`.

4m4n-x-B4w4ne replied:
Thanks ,I am working on it and also on clean PR build.

4m4n-x-B4w4ne replied:
I have wrote a basic-level test similar to other checks and also have done a clean PR build.
Can you please review it?

4m4n-x-B4w4ne replied:
I have written check file for all permutations of configuration of the options and also for the more comprehensive test cases as instructed with clean PR build.
Done Thanks.
@5chmidti  Can you please review it?

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Fixed Bugs regarding wrong variable names.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

Fixed Bugs regarding mismatched named variables.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

Improved Readibility and Description.

4m4n-x-B4w4ne committed:
Update implicit-bool-conversion-check.cpp

Fixed Bugs related to check-suffix.

4m4n-x-B4w4ne committed:
[Test]Update implicit-bool-conversion-check.cpp

Just for testing purpose of BuildKite.

4m4n-x-B4w4ne committed:
Update implicit-bool-conversion-check.cpp

Wrote basic level tests. (Similar to the tests in other check files).

4m4n-x-B4w4ne committed:
Update implicit-bool-conversion-check.cpp

Updated for all permutations of configurations options and also for  the more comprehensive tests.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -319,10 +319,11 @@ Changes in existing checks
+  <clang-tidy/checks/readability/implicit-bool-conversion>` check by adding the
+  option `UseUpperCaseLiteralSuffix` to select the case of the literal suffix in 
+  fixes and fixing false positive for implicit conversion of comparison result in 
+  C23 , and by adding the option `CheckConversionsToBool` or 
```
```suggestion
  C23, and by adding the option `CheckConversionsToBool` or 
```

4m4n-x-B4w4ne replied:
Done Thanks.

4m4n-x-B4w4ne replied:
Hey, What needs to be done further? Can you please help me out?

EugeneZelenko replied:
Please fix Clang-format complains.

4m4n-x-B4w4ne replied:
Done Thanks.

4m4n-x-B4w4ne committed:
Update ReleaseNotes.rst

HerrCai0907 reviewed (approved):
LGTM. If you do not have write access, please ping me. we can help you to do merge.

4m4n-x-B4w4ne committed:
Update ImplicitBoolConversionCheck.cpp

FIxed Clang formatting issue.

4m4n-x-B4w4ne commented:
@HerrCai0907 , Thanks I dont have any write permissions. Can you please help me to merge?

chapuni commented:
Excuse me, I've reverted this.

4m4n-x-B4w4ne commented:
> Excuse me, I've reverted this.

Can you help on what thing , It went wrong and like ,What shall I need to do to make it done?

kaddkaka commented:
I'm interested in this configuration. 

This was reverted after tests broke. What's the status on this now?

@chapuni do you have some inputs on how to do test verification before trying to merge this again?

chapuni commented:
@kaddkaka IIRC the issue was "char is unsigned". It could be reproducible with adding default target as failing targets above.
